### PR TITLE
Jw/webpack

### PR DESCRIPTION
**Pull request by jaredwy**
Focus on app.js
main.js
and of course webpack.config.js

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,11 +2,14 @@
...
 /out
 *.heapsnapshot
 etc/config/*.local.properties
 .travis-compilers
 /nbproject
 /package-lock.json
+*.bundle*.js
+*.map
+static/dist
+static/vs
```
Comment (RabsRincon) [inline .gitignore]: Minor: Newline

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -22,8 +22,7 @@ every commit, ensuring that they pass before commiting your changes. _You can di
 
 ## Basic code layout
 
-Code is separated into server-side code and client-side code. Server packages are installed via `package.json` and client-side via
-`bower.json`. Server code is in `app.js` and in the `lib` directory. Client code is all in the `static` directory.
+Code is separated into server-side code and client-side code. All dependencies(server and client side) are installed via `package.json`. Server code is in `app.js` and in the `lib` directory. Client code is all in the `static` directory.
```
Comment (RabsRincon) [inline CONTRIBUTING.md]: Space before `(`

File: Makefile [hunk 3]
```diff
@@ -1,6 +1,6 @@
 NODE_DIR?=/opt/compiler-explorer/node
-NPM:= $(shell env PATH="$(NODE_DIR)/bin:$$PATH" which npm)
-NODE:= $(shell env PATH="$(NODE_DIR)/bin:$$PATH" which node || env PATH="$(NODE_DIR)/bin:$$PATH" which nodejs)
+NPM:=$(shell env PATH=$(NODE_DIR)/bin:$(PATH) which yarn)
```
Comment (RabsRincon) [inline Makefile]: Is the change from $$ to $( right? If I recall correctly, we changed to $$ as per @apardoe's suggestion

File: app.js [hunk 4]
```diff
@@ -88,6 +91,17 @@ const archivedVersions = opts.archivedVersions;
 let gitReleaseName = "";
 let versionedRootPrefix = "";
 const wantedLanguage = opts.language || null;
+let builtResourcesRoot = "";
+
+
+const config = require('./webpack.config.js')[1],
```
Comment (RabsRincon) [inline app.js]: Maybe rename to webpackConfig?

File: app.js [hunk 5]
```diff
@@ -96,6 +110,10 @@ if (opts.static && fs.existsSync(opts.static + "/git_hash")) {
 }
 if (opts.static && fs.existsSync(opts.static + '/v/' + gitReleaseName))
     versionedRootPrefix = "v/" + gitReleaseName + "/";
+
+if(process.env.NODE_ENV == "DEV") {
```
Comment (RabsRincon) [inline app.js]: Unneeded type coercion. Minor: Space before `(`

File: app.js [hunk 6]
```diff
@@ -564,14 +582,31 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
...
         if (gitReleaseName) logger.info("  git release " + gitReleaseName);
 
         function renderConfig(extra) {
             const options = _.extend(extra, clientOptionsHandler.get());
             options.compilerExplorerOptions = JSON.stringify(options);
             options.root = versionedRootPrefix;
             options.extraBodyClass = extraBodyClass;
+            options.builtResourcesRoot = builtResourcesRoot;
+            options.require = function(path) { 
+                if(process.env.NODE_ENV == "DEV") {
```
Comment (RabsRincon) [inline app.js]: Unneeded type coercion

File: app.js [hunk 7]
```diff
@@ -564,14 +582,31 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
...
+                    //I have no idea why main => maps to styles i need to dig into this
+                    if(path == 'main.css') {
```
Comment (RabsRincon) [inline app.js]: Tyoe coercion

File: app.js [hunk 8]
```diff
@@ -564,14 +582,31 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
...
+                    //this will break assets in dev mode for now
+                    return '/dist/' + path;
+                }
+                if(staticManifest.hasOwnProperty(path)) {
+                    return versionedRootPrefix + "/dist/" + staticManifest[path];
+                }
+                if(assetManifest.hasOwnProperty(path)) {
+                    return versionedRootPrefix + "/dist/assets/" + assetManifest[path];
+                }
+                console.warn("requested an asset I don't know about");
```
Comment (RabsRincon) [inline app.js]: `logger.{warn,error}` better than console.log (Which will not let the tests pass)

File: app.js [hunk 9]
```diff
@@ -581,6 +616,21 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
             res.render('embed', renderConfig({embedded: true}));
         };
         const healthCheck = require('./lib/handlers/health-check');
+        
+        
+        if(process.env.NODE_ENV == "DEV") {
```
Comment (RabsRincon) [inline app.js]: Type coercion. Formatting

File: static/d-mode.js [hunk 10]
```diff
@@ -21,263 +21,259 @@
...
+}
 
-    monaco.languages.register({id: 'd'});
-    monaco.languages.setMonarchTokensProvider('d', definition());
-    monaco.languages.setLanguageConfiguration('d', configuration());
-});
\ No newline at end of file
+monaco.languages.register({id: 'd'});
+monaco.languages.setMonarchTokensProvider('d', definition());
+monaco.languages.setLanguageConfiguration('d', configuration());
```
Comment (RabsRincon) [inline static/d-mode.js]: Newline

File: static/monaco-loader.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+        window.require(['vs/editor/editor.main'], function() {
+            _loaded = true;
+            resolve(window.monaco);
+        });
+    });
+    return _loadPromise;
+};
+
+module.exports = waitForMonaco;
```
Comment (RabsRincon) [inline static/monaco-loader.js]: Newline

File: static/monaco.js [hunk 12]
```diff
@@ -23,9 +23,7 @@
...
 
-define(function (require) {
-    "use strict";
-    require('vs/editor/editor.main');
 
-    return monaco; // jshint ignore:line
-});
+
+//until webpack can support monaco properly it exists on the window
+module.exports = monaco;
```
Comment (RabsRincon) [inline static/monaco.js]: Newline

File: static/settings.js [hunk 13]
```diff
@@ -23,190 +23,188 @@
...
-        return setSettings;
-    }
+    });
+    $.data(themeSelect, 'last-theme', themeSelect.val());
+    return setSettings;
+}
 
-    return setupSettings;
-});
+module.exports = setupSettings;
```
Comment (RabsRincon) [inline static/settings.js]: Newline

File: webpack.config.js [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+const outputPathRelative = 'dist/';
+const staticRelative = 'static/';
+const staticPath = path.resolve(__dirname, staticRelative);
+const distPath = path.join(staticPath, outputPathRelative);
+const vsPath = path.join(staticPath, 'vs/');
+const assetPath = path.join(staticPath, "assets");
+const manifestPath = 'manifest.json';
+const outputname = process.env.NODE_ENV == "DEV" ? 'main.js' : 'bundle.[hash].js';
+const cssName = process.env.NODE_ENV  == "DEV" ? 'styles.css' :  "styles.[contenthash].css";
```
Comment (RabsRincon) [inline webpack.config.js]: Coercion here and above

File: webpack.config.js [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+const manifestPlugin = new ManifestPlugin({
+    fileName: 'manifest.json' //if you change this, you also need to update it in the app.js
```
Comment (RabsRincon) [inline webpack.config.js]: Shouldn't this be `fileName: manifestPath`?

File: Makefile (hunk 3 above)
Comment (jaredwy) [reply Makefile]: Ah, I just copied and pasted from one of the merges. Will fix that up.

**Comment by mattgodbolt** [comment]
Tweaked the `Makefile` a bit. We need to get `make test` / `npm test` working too. there's a precommit hook that @RabsRincon and I use (`make install-git-hooks`) that goes mental at the moment. For now I'm `commit -n`-ing though :)

**Comment by mattgodbolt** [comment]
Fixed githooks and initial dist thing - beta site is up, but needs some TLC yet.

File: Makefile [hunk 16]
```diff
@@ -1,6 +1,8 @@
 NODE_DIR?=/opt/compiler-explorer/node
-NPM:= $(shell env PATH="$(NODE_DIR)/bin:$$PATH" which npm)
-NODE:= $(shell env PATH="$(NODE_DIR)/bin:$$PATH" which node || env PATH="$(NODE_DIR)/bin:$$PATH" which nodejs)
+YARN_DIR?=/opt/compiler-explorer/yarn
+YARN_EXE:=$(shell env PATH=$(NODE_DIR)/bin:$(YARN_DIR)/bin:$(PATH) which yarn)
+NODE:=$(shell env PATH=$(NODE_DIR)/bin:$(PATH) which node || env PATH=$(NODE_DIR)/bin:$(PATH) which nodejs)
```
Comment (RabsRincon) [inline Makefile]: Todo: $$

File: app.js [hunk 17]
```diff
@@ -598,39 +651,23 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
...
-            .set('trust proxy', true)
-            .set('view engine', 'pug')
-            .use('/healthcheck', new healthCheck.HealthCheckHandler().handle) // before morgan so healthchecks aren't logged
-            .use(morgan('combined', {stream: logger.stream}))
-            .use(compression())
-            .use(router)
-            .use(Raven.errorHandler())
-            .on('error', err => logger.error('Caught error:', err, "(in web error handler; continuing)"));
-
-        const webAlias = ceProps("web-alias", "");
```
Comment (RabsRincon) [inline app.js]: @jaredwy is this now needed? How should this be implemented?

**Comment by jaredwy** [comment]
@RabsRincon i ahve opened an issue https://github.com/mattgodbolt/compiler-explorer/issues/802 to address the == issues at the root. Ill follow this Pr up with that (unless someone gets to it before me :D)

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -5,7 +5,8 @@ Before we go too far, an apology: Compiler Explorer grew out of a bit of hacky J
 well-used project pretty quickly. Not all the code was originally well-written or well-tested. Please be forgiving of that, 
 and be ready to help in improving that.
 
-Note that Compiler Explorer targets the latest Node.js LTS, so it's better if you do so as well when testing your changes locally. (But it will run in everything post-Node.js 8)
+Note that Compiler Explorer targets the latest Node.js LTS, so it's better if you do so as well when testing your changes locally.
+(But it *should* run in everything post-Node.js 8)
```
Comment (mattgodbolt) [inline CONTRIBUTING.md]: Haha :) good spot

File: app.js [hunk 19]
```diff
@@ -559,14 +572,25 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
...
         if (gitReleaseName) logger.info(`  git release ${gitReleaseName}`);
 
         function renderConfig(extra) {
             const options = _.extend(extra, clientOptionsHandler.get());
             options.compilerExplorerOptions = JSON.stringify(options);
-            options.root = versionedRootPrefix;
             options.extraBodyClass = extraBodyClass;
+            options.require = function (path) {
+                if (isDevMode()) {
+                    //this will break assets in dev mode for now
```
Comment (mattgodbolt) [inline app.js]: As discussed, we're totally cool with this for now :)

File: app.js [hunk 20]
```diff
@@ -559,14 +572,25 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
...
+                    return '/dist/' + path;
+                }
+                if(staticManifest.hasOwnProperty(path)) {
+                    return "dist/" + staticManifest[path];
+                }
+                if(assetManifest.hasOwnProperty(path)) {
+                    return "dist/assets/" + assetManifest[path];
+                }
+                logger.warn("Requested an asset I don't know about");
```
Comment (mattgodbolt) [inline app.js]: Nitpick: add the path here, else if we ever see it in the warnings we won't know what the missing path was :)

File: app.js [hunk 21]
```diff
@@ -575,8 +599,26 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
...
-        router
+        const healthCheck = require('./lib/handlers/health-check');
+
+        if (isDevMode()) {
+            webServer.use(webpackDevMiddleware(webpackCompiler, {
+                publicPath: webpackConfig.output.publicPath
+            }));
+            webServer.use(express.static(staticDir));
+        } else {
+            //assume that anything not dev is just production this gives sane defaults for anyone who isn't messing with this
```
Comment (mattgodbolt) [inline app.js]: :+1:

File: app.js (hunk 17 above)
Comment (mattgodbolt) [reply app.js]: I'm more than happy to fix this up after the merge: we're not using it (yet), but once we go outside of docker/nginx we'll need something like this to do beta.

File: lib/compilation-env.js [hunk 22]
```diff
@@ -101,9 +101,9 @@ CompilationEnvironment.prototype.enqueue = function (job) {
         try {
             return job();
         } catch (e) {
-            Raven.captureException(e);
-            logger.error("Caught promise exception " + e);
             if (e.stack) logger.info(e.stack);
+            logger.error("Caught promise exception " + e);
```
Comment (mattgodbolt) [inline lib/compilation-env.js]: Ah, nice catch

**Pull request merged**